Sources/SmithyCBOR/Reader/Reader.swift

```diff
@@ -0,0 +1,237 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import AwsCommonRuntimeKit
+import Foundation
+
+@_spi(SmithyReadWrite) import protocol SmithyReadWrite.SmithyReader
+@_spi(SmithyReadWrite) import enum SmithyReadWrite.ReaderError
+@_spi(SmithyReadWrite) import protocol SmithyReadWrite.SmithyWriter
+@_spi(Smithy) import struct Smithy.Document
+@_spi(Smithy) import protocol Smithy.SmithyDocument
+@_spi(SmithyTimestamps) import enum SmithyTimestamps.TimestampFormat
+@_spi(SmithyTimestamps) import struct SmithyTimestamps.TimestampFormatter
+
+@_spi(SmithyReadWrite)
+public final class Reader: SmithyReader {
+    public typealias NodeInfo = String
+
+    public let cborValue: CBORType?
+    public let nodeInfo: NodeInfo
+    public internal(set) var children: [Reader] = []
+    public internal(set) weak var parent: Reader?
+    public var hasContent: Bool { cborValue != nil && cborValue != .null }
+
+    public init(nodeInfo: NodeInfo, cborValue: CBORType?, parent: Reader? = nil) {
+        self.nodeInfo = nodeInfo
+        self.cborValue = cborValue
+        self.parent = parent
+        self.children = Self.children(from: cborValue, parent: self)
+    }
+
+    public static func from(data: Data) throws -> Reader {
+        let decoder = try CBORDecoder(data: [UInt8](data))
+        let rootValue: CBORType
+        if decoder.hasNext() {
+            rootValue = try decoder.popNext()
+        } else {
+            rootValue = .null
+        }
+        return Reader(nodeInfo: "", cborValue: rootValue, parent: nil)
+    }
+
+    private static func children(from cborValue: CBORType?, parent: Reader) -> [Reader] {
+        var children = [Reader]()
+        switch cborValue {
+        case .map(let map):
+            for (key, value) in map {
+                let child = Reader(nodeInfo: key, cborValue: value, parent: parent)
+                children.append(child)
+            }
+        case .array(let array):
+            for (index, value) in array.enumerated() {
+                let child = Reader(nodeInfo: "\(index)", cborValue: value, parent: parent)
+                children.append(child)
+            }
```

Comment:
Q:How are indefinite map and indefinite array CBOR values handled?

Reply:
I added handling for them to CRT Decoder and CRT Encoder in this PR: https://github.com/awslabs/aws-crt-swift/pull/305